azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,340 @@
+        """
+        Returns the package of the given Version found in the manifest. If not found, returns exception
+        """
+        for pkg in agent_manifest.pkg_list.versions:
+            if FlexibleVersion(pkg.version) == version:
+                # Found a matching package, only download that one
+                return pkg
+
+        raise Exception("No matching package found in the agent manifest for requested version: {0} in goal state incarnation: {1}, "
+                        "skipping agent update".format(str(version), self._gs_id))
+
+    @staticmethod
+    def __purge_extra_agents_from_disk(known_agents):
+        """
+        Remove from disk all directories and .zip files of unknown agents
+        (without removing the current, running agent).
+        """
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+
+        known_versions = [agent.version for agent in known_agents]
+        known_versions.append(CURRENT_VERSION)
+
+        for agent_path in glob.iglob(path):
+            try:
+                name = fileutil.trim_ext(agent_path, "zip")
+                m = AGENT_DIR_PATTERN.match(name)
+                if m is not None and FlexibleVersion(m.group(1)) not in known_versions:
+                    if os.path.isfile(agent_path):
+                        logger.info(u"Purging outdated Agent file {0}", agent_path)
+                        os.remove(agent_path)
+                    else:
+                        logger.info(u"Purging outdated Agent directory {0}", agent_path)
+                        shutil.rmtree(agent_path)
+            except Exception as e:
+                logger.warn(u"Purging {0} raised exception: {1}", agent_path, ustr(e))
+
+    @staticmethod
+    def __proceed_with_update(requested_version):
+        """
+        If requested version is specified, upgrade/downgrade to the specified version.
+        Raises: AgentUpgradeExitException
+        """
+        if requested_version < CURRENT_VERSION:
+            # In case of a downgrade, we mark the current agent as bad version to avoid starting it back up ever again
+            # (the expectation here being that if we get request to a downgrade,
+            # there's a good reason for not wanting the current version).
+            prefix = "downgrade"
+            try:
+                # We should always have an agent directory for the CURRENT_VERSION
+                agents_on_disk = AgentUpdateHandler.__get_available_agents_on_disk()
+                current_agent = next(agent for agent in agents_on_disk if agent.version == CURRENT_VERSION)
+                msg = "Marking the agent {0} as bad version since a downgrade was requested in the GoalState, " \
+                      "suggesting that we really don't want to execute any extensions using this version".format(CURRENT_VERSION)
+                logger.info(msg)
+                current_agent.mark_failure(is_fatal=True, reason=msg)
```

Comment:
The purge method should not delete blacklisted versions (it can delete the code, but leave the directory and error file on disk), otherwise the daemon may download and use them again